conans/client/tools/env.py

```diff
@@ -116,31 +118,65 @@ def remove_from_path(command):
 
 
 def env_diff(cmd, only_diff):
-    known_path_lists = ("include", "lib", "libpath", "path")
-    if platform.system() == "Windows":
-        cmd += " && set"
-    else:
-        cmd += " && export"
+    # portable variant of shell/export command, as export may have different results, e.g.:
+    # USER=sse4 (zsh)
+    # declare -x USER="sse4" (bash)
+    # or may not be available at all! (csh)
+    # output as dictionary, as values may contain spaces, newlines, etc.
+    uuid_str = uuid.uuid4().hex  # needed to reliably find the start of the script output
+    get_env = 'from __future__ import print_function; import os; import json; ' \
+              'print(\'%s\'); print(json.dumps(dict(os.environ)))' % uuid_str
+    known_path_lists = (
+        "INCLUDE",
+        "LIB",
+        "LIBPATH",
+        "PATH",
+        "CPATH",
+        "C_INCLUDE_PATH",
+        "CPLUS_INCLUDE_PATH",
+        "OBJC_INCLUDE_PATH",
+        "FPATH",
+        "MANPATH",
+        "PYTHONPATH",
+        "CLASSPATH",
+        "LD_LIBRARY_PATH",
+        "DYLD_LIBRARY_PATH",
+        "DYLD_FALLBACK_LIBRARY_PATH",
+        "DYLD_VERSIONED_LIBRARY_PATH",
+        "DYLD_FRAMEWORK_PATH",
+        "DYLD_FALLBACK_FRAMEWORK_PATH",
+        "DYLD_VERSIONED_FRAMEWORK_PATH",
+        "DYLD_ROOT_PATH",
+        "PERL5LIB",
+        "GCONV_PATH",
+        "LOCPATH",
+        "NIS_PATH,"
+        "NLSPATH",
+        "GOPATH"
+    )
+    cmd += " && \"%s\" -c \"%s\"" % (sys.executable, get_env)
```

Comment:
This approach is broken for the installers created with ``pyinstaller``:

https://pyinstaller.readthedocs.io/en/v3.6/runtime-information.html#using-sys-executable-and-sys-argv-0

Reply:
hmm, that's a good point.
in that case, we can't rely on `sys.executable`...
we also can't be sure there is `python` (or `python3`) executable in the PATH to run it.
what can we do in this case?
1. make code that works for all shells - seems to be hard and fragile (there are lots of them! - cmd.exe, powershell, zsh, bash, dash, csh, etc.)
2. execute conan (we know there is a conan executable, as we are running inside it, right?) with some special argument - sounds more like crutch and workaround, but should be reliable enough

any ideas?

Reply:
I think you are only left with the option of writing code for each individual shell if you don't assume you can execute python through `os.system`. You can't assume that there is a conan executable as conan didn't necessarily get run through a conan executable (e.g. python -m conans.conan). And even if there is a conan executable, if there is no python executable, it could very well be broken (after all, a conan executable is nothing more than a python script running conan).

Reply:
yeah, seems to be a vicious circle... let's try to write some reliable parser which satisfies all shells

Reply:
Can we assume that `/usr/bin/env` exists for unixy hosts? It at least is present on MacOS and every Linux distro that I'm aware of. It would be a consistent way of printing out environment variables.

Reply:
yes, seems to be the easiest way to achieve the goal - use existing code (vcvars) for Windows, and env for Linux/Mac/etc